core/loader_shared.c

```diff
@@ -991,11 +1006,18 @@ redirect_realloc(void *mem, size_t size)
 {
     void *buf = NULL;
     if (size > 0) {
+        /* XXX: This is not efficient at all.  For shrinking we should use the same
+         * object, and split off the rest for reuse: but DR's allocator was designed
+         * for DR's needs, and DR does not need that feature.  For now we do a
+         * very simple malloc+free.  In general, third-party lib code run by clients
+         * should be on slowpaths in any case.
+         */
         buf = redirect_malloc(size);
         if (buf != NULL && mem != NULL) {
-            size_t old_size = *((size_t *)((byte *)mem - sizeof(size_t)));
+            void *old_start;
+            size_t old_size = redirect_malloc_size_and_start(mem, &old_start);
             size_t min_size = MIN(old_size, size);
-            memcpy(buf, mem, min_size);
+            memcpy(buf, old_start, min_size);
```

Comment:
Why copy from old_start instead of mem? Didn't understand that part.

Reply:
Thank you, that is an error.  old_start is the DR alloc start, not the privlib start.  Wow, apparently I should not be the one writing this code: keep making errors.  At least my new test caught this one on Travis (passed locally somehow...b/c my 0xcd is DR's free pattern??  Confirmed: changing to 0x77 and it fails locally).

Reply:
This also applies to the size as well. AFAIK redirect_malloc_size_and_start() returns the alloc size no?

Reply:
Yes, should be redirect_malloc_requested_size().  Adding more to the test to catch that too.